common/unified/matrix/hybrid_kernels.cpp

```diff
@@ -0,0 +1,118 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/matrix/hybrid_kernels.hpp"
+
+
+#include <ginkgo/core/base/math.hpp>
+
+
+#include "common/unified/base/kernel_launch.hpp"
+#include "common/unified/base/kernel_launch_reduction.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace GKO_DEVICE_NAMESPACE {
+/**
+ * @brief The Hybrid matrix format namespace.
+ *
+ * @ingroup hybrid
+ */
+namespace hybrid {
+
+
+template <typename ValueType, typename IndexType>
+void convert_to_csr(std::shared_ptr<const DefaultExecutor> exec,
+                    const matrix::Hybrid<ValueType, IndexType>* source,
+                    const IndexType* ell_row_ptrs,
+                    const IndexType* coo_row_ptrs,
+                    matrix::Csr<ValueType, IndexType>* result)
+{
+    const auto ell = source->get_ell();
+    const auto coo = source->get_coo();
+    // ELL is stored in column-major, so we swap row and column parameters
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto ell_col, auto row, auto ell_stride, auto in_cols,
+                      auto in_vals, auto ell_row_ptrs, auto coo_row_ptrs,
+                      auto out_cols, auto out_vals) {
+            const auto ell_idx = ell_col * ell_stride + row;
+            const auto out_row_begin = ell_row_ptrs[row] + coo_row_ptrs[row];
+            const auto ell_row_size = ell_row_ptrs[row + 1] - ell_row_ptrs[row];
+            if (ell_col < ell_row_size) {
+                const auto out_idx = out_row_begin + ell_col;
+                out_cols[out_idx] = in_cols[ell_idx];
+                out_vals[out_idx] = in_vals[ell_idx];
+            }
+        },
+        dim<2>{ell->get_num_stored_elements_per_row(), ell->get_size()[0]},
+        static_cast<int64>(ell->get_stride()), ell->get_const_col_idxs(),
+        ell->get_const_values(), ell_row_ptrs, coo_row_ptrs,
+        result->get_col_idxs(), result->get_values());
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto idx, auto ell_row_ptrs, auto coo_row_ptrs,
+                      auto out_row_ptrs) {
+            out_row_ptrs[idx] = ell_row_ptrs[idx] + coo_row_ptrs[idx];
+        },
+        source->get_size()[0] + 1, ell_row_ptrs, coo_row_ptrs,
+        result->get_row_ptrs());
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto idx, auto in_rows, auto in_cols, auto in_vals,
+                      auto ell_row_ptrs, auto coo_row_ptrs, auto out_cols,
+                      auto out_vals) {
+            const auto row = in_rows[idx];
+            const auto col = in_cols[idx];
+            const auto val = in_vals[idx];
+            const auto coo_row_begin = coo_row_ptrs[row];
+            const auto coo_local_pos = idx - coo_row_begin;
+            // compute row_ptrs[row] + ell_row_size[row]
```

Comment:
```suggestion
```

Reply:
This is a comment trying to explain the following, convoluted way of computing the output index.